websocket.go

```diff
@@ -0,0 +1,359 @@
+			log.Printf("Failed to forward unsubscribe message to upstream: %v", err)
+		}
+	}
+}
+
+func (p *WebSocketProxy) Start() {
+	// Connect to upstream WebSocket server
+	err := p.connectToUpstream()
+	if err != nil {
+		log.Fatalf("Failed to connect to upstream WebSocket: %v", err)
+	}
+
+	// Start a goroutine to listen for messages from the upstream server
+	go func() {
+		for {
+			_, message, err := p.upstream.Read(context.Background())
+			if err != nil {
+				log.Printf("Upstream WebSocket read error: %v", err)
+
+				// Attempt to reconnect to the upstream server
+				for {
+					log.Println("Attempting to reconnect to upstream server...")
+					err := p.connectToUpstream()
+					if err == nil {
+						log.Println("Reconnected to upstream server successfully")
+
+						// Resubscribe to all swap IDs after reconnection
+						p.mu.Lock()
+						swapIDs := make([]string, 0, len(p.subscribers))
+						for swapID := range p.subscribers {
+							swapIDs = append(swapIDs, swapID)
+						}
+						p.mu.Unlock()
+
+						if len(swapIDs) > 0 {
+							resubscribeMsg := map[string]any{
+								"op":      "subscribe",
+								"channel": "swap.update",
+								"args":    swapIDs,
+							}
+							messageBytes, err := json.Marshal(resubscribeMsg)
+							if err != nil {
+								log.Printf("Failed to marshal resubscribe message: %v", err)
+							} else {
+								err = p.upstream.Write(context.Background(), websocket.MessageText, messageBytes)
+								if err != nil {
+									log.Printf("Failed to forward resubscribe message to upstream: %v", err)
+								} else {
+									log.Printf("Resubscribed to swap IDs: %v", swapIDs)
+								}
+							}
+						}
+
+						break
+					}
+					log.Printf("Failed to reconnect to upstream server: %v", err)
+
+					// Wait before retrying
+					time.Sleep(5 * time.Second)
+				}
+
+				// Continue the loop to resume processing messages
+				continue
+			}
+
+			// Parse the message
+			var msg map[string]any
+			if err := json.Unmarshal(message, &msg); err != nil {
+				log.Printf("Failed to parse upstream message: %v", err)
+				continue
+			}
+
+			// Extract swap IDs from the message
+			args, ok := msg["args"].([]any)
+			if !ok {
+				log.Printf("Invalid upstream message: %v", msg)
+				continue
+			}
+
+			// Group updates by client
+			clientUpdates := make(map[*websocket.Conn][]map[string]any)
+			p.mu.Lock()
+			for _, arg := range args {
+				swap, ok := arg.(map[string]any)
+				if !ok {
+					continue
+				}
+				swapID, ok := swap["id"].(string)
+				if !ok {
+					continue
+				}
+
+				// Add updates to each subscribed client
+				for client := range p.subscribers[swapID] {
+					clientUpdates[client] = append(clientUpdates[client], swap)
+				}
+			}
+			p.mu.Unlock() // Release the lock after grouping updates
+
+			// Notify clients
+			for client, updates := range clientUpdates {
+				// Create a single notification message for the client
+				notification := map[string]any{
```

Comment:
Did you consider taking the message as is from the upstream and sent it to the subscribers in stead of creating the notification object?